### ISSUE TITLE
Database is locked

### ISSUE THREAD

**Issue by mobamoba** [OfflineIMAP/offlineimap#401]
Since upgrading to 7.0.9 I keep getting the following error:
2016-11-05 11:35:29 WARNING: ERROR: Exceptions occurred during the run!
2016-11-05 11:35:29 WARNING: ERROR: ERROR in syncfolder for Exchange folder INBOX: Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/accounts.py", line 536, in syncfolder
    statusfolder.openfiles()
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/LocalStatusSQLite.py", line 135, in openfiles
    self.__create_db()
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/LocalStatusSQLite.py", line 232, in __create_db
    """)
OperationalError: database is locked

  database is locked

I tried rebooting in the hopes that the database would unlock itself but that didn't help. Any thoughts on how to resolve this issue? Thanks.

**Comment by nicolas33** [comment]
Please, provide all the information requested at issue creation time. Create a new issue if this one is closed or locked.

Otherwise, it's not possible to correctly dig into this issue.

Also, what was your previous offlineimap version?

**Comment by mobamoba** [comment]
Sorry about that - I mean to copy,paste it and forgot. I was using this version (https://github.com/OfflineIMAP/offlineimap/issues/380)  then updated to 7.0.9.

OI started screwing up again. I would archive a message from my Inbox and OI would readd it. I ended up with like 10 copies of the same message and OI kept throwing an error about UNIQUE constraint failed in my archive folder along with the database locked error I listed above.

I then stopped OI and followed the Folder Validity instructions in the FAQ:
rm -r ~/.offlineimap/Account-AccountName/LocalStatus/INBOX
rm -r ~/.offlineimap/Repository-RemoteRepositoryName/FolderValidity/INBOX

This got rid of the UNIQUE constraint failed issue but left me with the database locked issue unresolved.



General informations

Ubuntu 16.04:
offlineimap version 7.09
Python version: 2.7
server name or domain:
CLI options:
Configuration file offlineimaprc

[general]
accounts = 1, 2, 3, 4, 5, 6
maxsyncaccounts = 7
ignore-readonly = no
socktimeout = 60

[Account Exchange]
localrepository = Local
remoterepository = Remote
autorefresh = -1
maxconnections = 2
status_backend = sqlite

[Repository Local]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com
nametrans = lambda sentname: re.sub ('Sent Items', 'Sent Current',
                sentname)

[Repository Remote]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
nametrans = lambda sentname: re.sub ('Sent Current', 'Sent Items',
                sentname)
folderfilter = lambda foldername: foldername in [
    'INBOX', 'Saved Current', 'Sent Current', 'Drafts', 'Trash', '[Gmail]/Spam',
    'SMS', 'SMSArchived', 'Call log', 'Call log Archived']
keepalive = 60

[Account 1]
localrepository = Local-1
remoterepository = Remote-1
autorefresh = 20
maxconnections = 1
status_backend = sqlite

[Repository Local-1]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com

[Repository Remote-1]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
folderfilter = lambda foldername: foldername in [
    'SLISent']
nametrans = lambda sentname: re.sub ('SLISent', 'Sent SLI',
                sentname)
readonly = True

[Account 2]
localrepository = Local-2
remoterepository = Remote-2
autorefresh = 20
maxconnections = 1
status_backend = sqlite

[Repository Local-2]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com

[Repository Remote-2]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
folderfilter = lambda foldername: foldername in [
    '2Sent']
nametrans = lambda sentname: re.sub ('2Sent', 'Sent 2',
                sentname)
readonly = True

[Account 3]
localrepository = Local-3
remoterepository = Remote-3
autorefresh = 20
maxconnections = 1
status_backend = sqlite

[Repository Local-3]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com

[Repository Remote-3]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
folderfilter = lambda foldername: foldername in [
    '3Sent']
nametrans = lambda sentname: re.sub ('3Sent', 'Sent 3',
                sentname)
readonly = True

[Account 4]
localrepository = Local-4
remoterepository = Remote-4
autorefresh = 20
maxconnections = 1
status_backend = sqlite

[Repository Local-4]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com

[Repository Remote-4]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
folderfilter = lambda foldername: foldername in [
    '4Sent']
nametrans = lambda sentname: re.sub ('4Sent', 'Sent 4',
                sentname)
readonly = True

[Account 5]
localrepository = Local-5
remoterepository = Remote-5
autorefresh = 20
maxconnections = 1
status_backend = sqlite

[Repository Local-5]
type = IMAP
remotehost = www.---.com
ssl = yes
cert_fingerprint = ---
remotepass = ---
remoteuser = ---@---.com

[Repository Remote-5]
type = Gmail
remotepass = ---
remoteuser = ---@---.com
sslcacertfile = ---.crt
folderfilter = lambda foldername: foldername in [
    '5Sent']
nametrans = lambda sentname: re.sub ('5Sent', 'Sent 5',
                sentname)
readonly = True```

#### pythonfile (if any)

**Comment by nicolas33** [comment]
> I would archive a message from my Inbox

* What do you mean by "archive" a message? What do you do exactly?
* Is "Account Exchange" the failing account?
* Does v7.0.8 works?

**Comment by mobamoba** [comment]
By archive I simply mean move it to a different folder. Account Exchange is the failing account but only in the Inbox. 
I never saw either or those errors on 7.0.8. I haven't tried reinstalling it (I'm not even sure I have it as I simply clone the git master and make OI from that).

**Comment by nicolas33** [comment]
Great. It should be possible to move to your local repository and `git checkout v7.0.8`.
Let us know if you still have this issue.

**Comment by mobamoba** [comment]
Same database locked error in 7.0.8.

**Comment by nicolas33** [comment]
Is this reproducible with `maxconnections = 1`?

**Comment by mobamoba** [comment]
Yes. That setting didn't solve the problem.

**Comment by nicolas33** [comment]
Oh, `autorefresh` is set to **-1**! IOW, I think the next sync starts before the current run is finished. Negative values should be forbidden, here.

**Comment by mobamoba** [comment]
I like the constant autorefresh so please don't change that option! But regardless it's not the problem because I get the same issue even when I set OI with -o to run only once.

**Comment by nicolas33** [comment]
I'll change it to forbid negative values anyway. It's just plain wrong and might cause serious damage.

**Comment by nicolas33** [comment]
Side note: I'm pretty sure the previous truncated UIDMaps was due to the negative value of autorefresh.

**Comment by nicolas33** [comment]
First, checkout the latest v7.0.9.

What the following commands give?
```
$ fuser <metadata>/Account-<name>/LocalStatus-sqlite/INBOX
or
$ lsof <metadata>/Account-<name>/LocalStatus-sqlite/INBOX

$ sqlite3 --version
```

Also, please try above patch.

As the last resort, try:
```
$ cd <metadata>/Account-<name>/LocalStatus-sqlite/
$ cp INBOX INBOX.bak
$ echo '.dump' | INBOX.bak | INBOX
```

**Comment by mobamoba** [comment]
The patch didn't fix anything. Fuser and lsof did nothing. Sqlite is 3.11.0. there must be something off in the syntax of the echo command because I got INBOX command not found and INBOX.bak command not found.

**Comment by nicolas33** [comment]
I wonder you replaced the placeholders by real values. Can't you find the databases in the metadata path?

**Comment by mobamoba** [comment]
Yeah I filled in my actual folder and account name.

**Comment by nicolas33** [comment]
My bad. It should be `$ echo '.dump' | sqlite3 INBOX.bak | sqlite3 INBOX`, indeed.

**Comment by mobamoba** [comment]
I was able to execute the command, but same error - database locked.

**Comment by nicolas33** [comment]
I can't get why this fails and I cannot reproduce the bug. Perhaps you could git bisect it:
http://www.offlineimap.org/doc/git-bisect.html

**Comment by mobamoba** [comment]
I can try that but first: do you think there's some kind of lock file somewhere that's preventing this from working? In other words, could it be that sqlite or python or something incorrectly thinks this file is locked because something that should've have been deleted - a lock file, a cache, something - is still sticking around?

**Comment by nicolas33** [comment]
> do you think there's some kind of lock file somewhere that's preventing this from working? In other words, could it be that sqlite or python or something incorrectly thinks this file is locked because something that should've have been deleted - a lock file, a cache, something - is still sticking around?

I can say for sure that no other process is accessing the database. At least, that's what the empty outputs of fuser/lsof are saying.

We can still wonder a corrupted database. It's possible to rule out this possibility by inserting/removing a new line (offlineimap NOT running):
```
$ sqlite3 INBOX
sqlite> select * from status;
sqlite> INSERT INTO STATUS (id,flags,mtime,labels) VALUES (1, "S", 0, "");
sqlite> DELETE FROM status where id=1;
sqlite> .quit
```

To check against Python and sqlite3 module bugs we'd have to write a small python script.

**Comment by mobamoba** [comment]
sqlite> select * from status; gave me "Error: no such table: status"

**Comment by nicolas33** [comment]
`echo ".schema" | sqlite3 INBOX`

**Comment by mobamoba** [comment]
that returned nothing.

**Comment by nicolas33** [comment]
And with the INBOX.bak ?

**Comment by mobamoba** [comment]
Same thing.

**Comment by nicolas33** [comment]
Remove those files!

**Comment by mobamoba** [comment]
Nope that did work at all. Here's my debug log (offlineimap -o -a Exchange -d imap -l oi3debug.log). I know it's long but you can see the errors.

Oh - and perhaps this is a new thread - it would be great if the debug log automatically deleted the XOAUTH2 secret, key, etc. like it does with passwords because I had to manually go through and delete them from the file since they're obviously not something I want to share publicly.

[oi3debug.txt](https://github.com/OfflineIMAP/offlineimap/files/581429/oi3debug.txt)

**Comment by nicolas33** [comment]
I don't see anymore "Database is locked" for v7.0.9 in this file.

**Comment by mobamoba** [comment]
Search for error - it's the first one that comes up.

**Comment by nicolas33** [comment]
I won't check the log for v7.0.8 if it's fixed with v7.0.9. It does not make sense to fix issues already fixed.

**Comment by mobamoba** [comment]
Sorry about that. I was using 7.0.9 but the log file was appending (meaning the latest log was at the bottom). It made no difference - same two errors:
Error 1:
2016-11-09 12:47:21 INFO: Creating new Local Status db for Local:INBOX
2016-11-09 12:47:26 ERROR: ERROR: ERROR in syncfolder for Exchange folder INBOX: Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/accounts.py", line 536, in syncfolder
    statusfolder.openfiles()
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/LocalStatusSQLite.py", line 135, in openfiles
    self.__create_db()
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/LocalStatusSQLite.py", line 232, in __create_db
    """)
OperationalError: database is locked

  database is locked

Error two - this error comes up multiple times with different folders but is the same error every time:

2016-11-09 12:49:03 ERROR: ERROR: while syncing Saved Current [account Exchange]
  UNIQUE constraint failed: status.id
2016-11-09 12:49:03 ERROR: ['  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/Base.py", line 1110, in syncmessagesto\n    action(dstfolder, statusfolder)\n', '  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/Base.py", line 922, in __syncmessagesto_copy\n    statusfolder.savemessage(uid, None, flags, rtime)\n', '  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/LocalStatusSQLite.py", line 356, in savemessage\n    (uid,flags,mtime,labels))\n', '  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/LocalStatusSQLite.py", line 184, in __sql_write\n    self.connection.execute(sql, args)\n']
2016-11-09 12:49:03 ERROR: ERROR: ERROR in syncfolder for Exchange folder Saved Current: Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/accounts.py", line 584, in syncfolder
    remotefolder.syncmessagesto(localfolder, statusfolder)
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/Base.py", line 1110, in syncmessagesto
    action(dstfolder, statusfolder)
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/Base.py", line 922, in __syncmessagesto_copy
    statusfolder.savemessage(uid, None, flags, rtime)
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/LocalStatusSQLite.py", line 356, in savemessage
    (uid,flags,mtime,labels))
  File "/usr/local/lib/python2.7/dist-packages/offlineimap/folder/LocalStatusSQLite.py", line 184, in __sql_write
    self.connection.execute(sql, args)
IntegrityError: UNIQUE constraint failed: status.id

  UNIQUE constraint failed: status.id

**Comment by nicolas33** [comment]
Interesting. It appears that there is another error:

`ERROR: while syncing Saved Current [account Exchange]
UNIQUE constraint failed: status.id`

This clearly show that there are concurrency issues ind offlineimap. Since I don't hit this issue I guess this comes from the IMAP/IMAP use case only.

Do you still get this issue with `singlethreadperfolder = yes` in the remote section?

**Comment by nicolas33** [comment]
I get what's happening. The internal database lock is not shared by all the objects as expected.

Setting the singlethreadperfolder should workaround the issue while this is not fixed.

**Comment by nicolas33** [comment]
> I get what's happening. The internal database lock is not shared by all the objects as expected.

My bad. The lock looks shared correctly. This is a hint that I'm too tired. I'll dig into this another day.

I'm still curious about the singlethreadperfolder test, though.

**Comment by nicolas33** [comment]
At this point, I'd git bisect too.

**Comment by mobamoba** [comment]
I can git bisect but we already know (from that appended debug log I put up the other day) that the problem was happening when I reinstalled 7.0.8. However, when I was actually using 7.0.8 a few months ago (as opposed to making a new one from git) I didn't have the problem. Therefore I'm wondering if something got corrupted in the sqlite files somewhere and no amount of bisecting will fix it because the problem isn't with OI's code. See what I mean?

**Comment by nicolas33** [comment]
Yes. I'd restart from scratch. Move everything about the cache out, empty one side and sync again.

**Comment by mobamoba** [comment]
Is there a way to do that without ending up with thousands of duplicate emails? Like can OI record the run for database purposes without actually downloading or uploading any emails?

**Comment by nicolas33** [comment]
No, it's not possible. That's why one side must be empty.

Here's the obvious safe way:
1. Create a new account to save the remote to a local maildir.
2. Empty this remote.
3. IMAP/IMAP sync (after metadata removal).

If required (emails missing, whatever):
4. Create another new account to redownload to a another local maildir.
5. Copy the **maildir** made at step 1 to the local maildir made at step 4.
6. Sync account created at step 4.

**Comment by mobamoba** [comment]
That's rough. And ultimately it doesn't really tell us anything in terms of debugging does it? I mean I could go through all that and wind up with the problem again because we don't know what caused it. Is there any way to do further debugging to figure out the underlying sqlite issue?

**Comment by nicolas33** [comment]
This would tell us if there is any issue, actually. For now, you're the only one having hit such troubles.

**Comment by mobamoba** [comment]
So I tried a very simple experiment, which was to try this on one folder. I deleted the INBOX file from Account-Exchange and Repository-Local (it didn't exist in Repository-Remote) and all subfolders. Same error. Does this tell you anything?

**Comment by nicolas33** [comment]
None response for the [above request](https://github.com/OfflineIMAP/offlineimap/issues/401#issuecomment-259828282), closing.

**Comment by nicolas33** [comment]
@jwillhoeft Good feedback, very appreciated. Thank you!
